src/nb_type.cpp

```diff
@@ -324,6 +386,26 @@ PyObject *nb_type_new(const type_data *t) noexcept {
     type->tp_init = inst_init;
     type->tp_new = inst_new;
 
+    // If our base class has dynamic attributes, we must enable them, too, to
+    // match the base's layout.
+    // enable_dynamic_attributes is harmless to call if the base class
+    // has dynamic attributes also enabled.
+    if (has_dynamic_attr || (base && base->tp_dictoffset))
+    {
+        // We're about to set our tp_dictoffset to this value.
+        // The base class must be consistent with that.
+        assert(!base
+                || !base->tp_dictoffset
+                || (base->tp_dictoffset == type->tp_basicsize));
+
+        enable_dynamic_attributes(&nbt->ht);
+    }
+
+    // tp_basicsize is computed above, depending on whether the class has
+    // a dict pointer for dynamic attributes. tp_basicsize can only
+    // grow, not shrink.
+    assert(!has_base || (base->tp_basicsize <= type->tp_basicsize));
```

Comment:
Please take a close look here. I'm using the presence of `base->tp_dictoffset` to decide whether the base class has dynamic attributes and act accordingly. As long as it's C++ types only, I think that's safe. I've tried to be reasonably paranoid, but I don't understand where `base_py` might come from and how it might influence what needs to happen.